metsim/methods/mtclim.py

```diff
@@ -106,127 +88,90 @@ def _calc_tfmax(precip, dtr, sm_dtr):
-        parray = np.zeros(params['n_days'])
-        start_yday = df['day_of_year'][0]
-        end_yday = df['day_of_year'][-1]
-
+        window = np.zeros(params['n_days'] + 90)
+        window[90:] = df['precip']
+        
         # If yeardays at end match with those at beginning we can use
         # the end of the input to generate the beginning by looping around
         # If not, just duplicate the first 90 days
-        if start_yday != 1:
-            if end_yday == start_yday-1:
-                isloop = True
-        else:
-            if end_yday == 365 or end_yday == 366:
-                isloop = True
-        if isloop:
-            for i in range(90):
-                window[i] = df['precip'][params['n_days'] - 90 + i]
+        start_day, end_day = df['day_of_year'][0], df['day_of_year'][-1]
+        if (start_day%365 == (end_day%365)+1) or (start_day%366 == (end_day%366)+1):
+            window[:90] = df['precip'][-90:]
         else:
-            for i in range(90):
-                window[i] = df['precip'][i]
-        window[90:] = df['precip']
+            window[:90] = df['precip'][:90]
 
-        # FIXME: This can probably be vectorized
-        for i in range(params['n_days']):
-            sum_precip = 0.0
-            for j in range(90):
-                sum_precip += window[i+j]
-                sum_precip = (sum_precip / 90.) * consts['DAYS_PER_YEAR']
-            sum_precip = np.maximum(sum_precip, 8.0)
-            parray[i] = sum_precip
+        parray = np.array(pd.Series(window)
+                            .rolling(window=90, win_type=win_type,axis=0)
+                            .mean())[90:] * cnst.DAYS_PER_YEAR 
+    # Convert to mm 
+    parray = np.maximum(parray, 80.0) / 10
 
     df['tfmax'] = _calc_tfmax(df['precip'], dtr, sm_dtr) 
-
     tdew = df.get('tdew', df['t_min'])
     pva = df.get('hum', svp(tdew))
-    pa = atm_pres(params['site_elev'])
+    pa = atm_pres(df['elev'][0])
+    #pa = atm_pres(5688)
     yday = df['day_of_year'] - 1 
     df['dayl'] = sg['daylength'][yday]
-
+ 
+    # Calculation of tdew and swrad. tdew is iterated on until
+    # it converges sufficiently 
+    tdew_old = tdew
+    tdew, pva = sw_hum_iter(df, sg, pa, pva, parray, dtr)
+    while(np.sqrt(np.mean((tdew-tdew_old)**2)) > 1e-3):
+        tdew_old = np.copy(tdew)
+        tdew, pva = sw_hum_iter(df, sg, pa, pva, parray, dtr)
+    df['vapor_pressure'] = pva 
+
+
+def sw_hum_iter(df, sg, pa, pva, parray, dtr):
     tt_max0 = sg['tt_max0']
     potrad = sg['potrad']
     daylength = sg['daylength']
     yday = df['day_of_year'] - 1
 
-    # Calculation of shortwave is split into 3 components:
-    # 1. Direct radiation arriving from incident light
-    # 2. Diffuse radiation over entire daylenght
-    # 3. Influence of snowpack - optionally set by MTCLIM_SWE_CORR
-    t_tmax = np.maximum(tt_max0[yday] + (params['ABASE'] * pva), 0.0001)
+    t_tmax = np.maximum(tt_max0[yday] + (cnst.ABASE * pva), 0.0001)
```

Comment:
`solar_geom` unfortunately can't be turned into a Pandas dataframe due to the structure of `tiny_rad_fract`.  Since that's the case we don't have anything to call `.where` on.